DROOLS-4118: [DMN Designer] Documentation tab - Show information from the DMN diagram in the documentation tab

karreiro opened the pull request:
See:
- https://issues.jboss.org/browse/DROOLS-4117
- https://issues.jboss.org/browse/DROOLS-4118

---

**Demo:**

![2019-06-11 18 24 15](https://user-images.githubusercontent.com/1079279/59311374-9af51900-8c7f-11e9-8f5a-82202eda28a2.gif)

---

@tiagodolphine I've applied the use of  `PathHelper` on stunner (`BPMNDocumentationView.java`). I just tried the **Documentation** tab there and everything looks fine, but I'm adding you for double-checking it ;-) (thus, feel free to just review stunner related classes):

![2019-06-11 19 29 35](https://user-images.githubusercontent.com/1079279/59311425-d98ad380-8c7f-11e9-8a3e-6de124877be4.gif)

---

Part of an ensemble:
- https://github.com/kiegroup/droolsjbpm-build-bootstrap/pull/971
- https://github.com/kiegroup/kie-wb-common/pull/2694

karreiro commented:
Jenkins, please retest this.

karreiro commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationService.java:
```diff
@@ -0,0 +1,30 @@
+import org.kie.workbench.common.stunner.core.documentation.service.DiagramDocumentationService;
+
+public interface DMNDocumentationService extends DiagramDocumentationService<
+        Diagram,
```
Thank you, Mike :-) Fixed!

jomarko commented:
### Manual Review - Note#01
The selected nodes are highlighted also in documentation, should I report this as an issue?

![Screenshot from 2019-06-13 11-31-38](https://user-images.githubusercontent.com/8044780/59421406-d5b99700-8dce-11e9-8b7b-21e4e2be4e6c.png)

jomarko commented:
### Manaul Review - Note#02
Background colors are not in documentation. Changed Font is.
![Screenshot from 2019-06-13 11-34-24](https://user-images.githubusercontent.com/8044780/59421685-4eb8ee80-8dcf-11e9-99ce-a65e99830651.png)

![Screenshot from 2019-06-13 11-34-29](https://user-images.githubusercontent.com/8044780/59421694-524c7580-8dcf-11e9-9d15-a75cc2aedcfa.png)

jomarko commented:
### Manual Review - Note#03
The print button looks different for BPMN and DMN

DMN
![Screenshot from 2019-06-13 11-36-04](https://user-images.githubusercontent.com/8044780/59421838-92135d00-8dcf-11e9-80c1-04952db4c8c2.png)

BPMN
![Screenshot from 2019-06-13 11-36-11](https://user-images.githubusercontent.com/8044780/59421852-9a6b9800-8dcf-11e9-9e4d-abc6ad65928c.png)

jomarko commented:
Manual review finished

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java:
```diff
@@ -0,0 +1,189 @@
+    }
+
+    String getFileName(final Diagram diagram) {
+        Metadata metadata = diagram.getMetadata();
```
can be also `final`

karreiro replied:
Fixed! Thanks :-)

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java:
```diff
@@ -0,0 +1,189 @@
+    private String getConstraint(final ItemDefinition itemDefinition) {
+        final UnaryTests allowedValues = itemDefinition.getAllowedValues();
+
+        if (allowedValues != null && !allowedValues.getText().getValue().isEmpty()) {
```
Could we please use `StringUtils` ?

karreiro replied:
Fixed! Thanks :-)

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/java/org/kie/workbench/common/dmn/client/editors/documentation/common/DMNDocumentationFactory.java:
```diff
@@ -0,0 +1,189 @@
+        return sessionInfo.getIdentity().getIdentifier();
+    }
+
+    private Optional<AbstractCanvasHandler> getCanvasHandler() {
```
I am not sure if this isn't already present in `DMNGraphUtils.getCanvasHandler`

karreiro replied:
Fixed! Thanks :-)

jomarko commented on kie-wb-common-dmn/kie-wb-common-dmn-client/src/main/resources/org/kie/workbench/common/dmn/client/resources/i18n/DMNClientConstants.properties:
```diff
@@ -245,3 +245,18 @@ DecisionComponentsView.EmptyStateDescription=To add an external components to th
+DMNDocumentationI18n.DiagramDoesNotHaveElements=The diagram does not have any element.
+DMNDocumentationFactory.NoFileName=No file name
+DMNDocumentationFactory.Constraints=Constraints :
+DMNDocumentationFactory.ListYes=List: Yes
```
Slightly inconsistency I think:
```
Constraints :
List : yes
```

vs

```
Constraints:
List: yes
```

karreiro replied:
Fixed! Thanks :-)

tiagodolphine commented on kie-wb-common-stunner/kie-wb-common-stunner-sets/kie-wb-common-stunner-bpmn/kie-wb-common-stunner-bpmn-client/src/main/java/org/kie/workbench/common/stunner/bpmn/client/documentation/BPMNDocumentationView.java:
```diff
@@ -92,61 +99,8 @@ public BPMNDocumentationView initialize(Diagram diagram) {
-    private void print() {
-        print(documentationDiv);
+    void print() {
+        printHelper.print(documentationDiv);
```
It looks good to me the use of the printHelper here, in this way the print logic in centralized there using elmental2, good change. Thanks.

karreiro replied:
Thank you, @tiagodolphine ☺️

karreiro commented:
Thank you for the review, @jomarko. Changes applied.

---

> **Manual Review - Note#01**
> The selected nodes are highlighted also in documentation, should I report this as an issue?

Since this behaviour happens in the Stunner editor, and every place the exports a canvas, I think that we can handle it in a separated JIRA.


> **Manaul Review - Note#02**
> Background colors are not in documentation. Changed Font is.

Fixed. Good catch!

![2019-06-13 14 01 45](https://user-images.githubusercontent.com/1079279/59453004-584d5100-8de5-11e9-9a3c-749570aa90ff.gif)


> **Manual Review - Note#03**
> The print button looks different for BPMN and DMN

I don't think this is a problem. I've checked the BC and seems like a standard to use the icon in the left side, see:
<img width="620" alt="Screen Shot 2019-06-13 at 11 09 33" src="https://user-images.githubusercontent.com/1079279/59453160-abbf9f00-8de5-11e9-8977-4bf0bd37cc9e.png">
<img width="676" alt="Screen Shot 2019-06-13 at 11 07 00" src="https://user-images.githubusercontent.com/1079279/59453161-abbf9f00-8de5-11e9-8734-eaab05de7178.png">
<img width="596" alt="Screen Shot 2019-06-13 at 11 06 42" src="https://user-images.githubusercontent.com/1079279/59453162-ac583580-8de5-11e9-9b45-11fcadca9b3f.png">
<img width="1423" alt="Screen Shot 2019-06-13 at 11 05 09" src="https://user-images.githubusercontent.com/1079279/59453163-ac583580-8de5-11e9-8558-ac96e6fae553.png">

Thus, I think we can keep as is for now, and check in the Monday meeting if the button is fine like this. If not, we can fix it in a separated JIRA.

Thanks again for the review :-)

karreiro commented:
Jenkins please retest this.

jomarko commented:
The issue for **Manual Review Note#01** reported here https://issues.jboss.org/browse/DROOLS-4178

karreiro commented:
Jenkins please retest this.

karreiro commented:
Jenkins please retest this.

karreiro commented:
Jenkins please retest this 😥

karreiro commented:
Jenkins please retest this.

karreiro commented:
Jenkins please retest this.

cristianonicolai commented:
jenkins please retest this

karreiro commented:
(rebasing the branch...) 👀

karreiro commented:
Jenkins please retest this.

karreiro commented:
Jenkins please retest this.

karreiro commented:
(fixing conflicts...) 👀

karreiro commented:
Jenkins please retest this.

karreiro commented:
Jenkins please retest this.

karreiro commented:
Jenkins please retest this.

karreiro commented:
@tiagodolphine Moving `PrintHelper` from (widgets):
`org.kie.workbench.common.widgets.client.util.PrintHelper;`
to (stunner):
`org.kie.workbench.common.stunner.core.client.util.PrintHelper;`

tiagodolphine commented:
Ok @karreiro LGTM
Thanks.

karreiro commented:
Jenkins please retest this.
